From 4a4012fc904abd20294eea55e9017fa53e48aec9 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 7 Mar 2025 10:20:38 +0100
Subject: [PATCH 1632/1695] clk: stm32mp25: Check Peripheral RIF assignation
 before configuring the MUX

Some peripherals have a MUX (e.g., ADCx, USBxPHYx) and can be configured
for the M33.
After a standby, the Cortex A35 has to check this configuration to avoid
an IAC.
It is the responsibility of the M33 to restore this MUX.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 63e9a230872fd2fdb59599a8f2c11403b3d094cc clk: stm32mp25: Check Peripheral RIF assignation before configuring the MUX]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I263cfcb4390ee79501fec8be32d357b40b46dc4f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/442716
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 1bc36df25..b6524b967 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -2274,15 +2274,20 @@ static int stm32_clk_configure_mux(struct clk_stm32_priv *priv __unused,
 	bool sem_taken = false;
 	int ret = 0;
 
-	if (tab_mux_rifsc[mux].is_rifsc &&
-	    stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1)) {
-		if (stm32_rif_acquire_semaphore(rifsc_base +
-						_RIFSC_RISC_PER0_SEMCR +
-						per_offset,
-						MAX_CID_SUPPORTED))
-			return -1;
-
-		sem_taken = true;
+	if (tab_mux_rifsc[mux].is_rifsc) {
+		if ((cidcfgr & _CIDCFGR_CFEN) &&
+		    !stm32_rif_scid_ok(cidcfgr, _CIDCRGR_SCID_MASK, RIF_CID1))
+			return 0;
+
+		if (stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1)) {
+			if (stm32_rif_acquire_semaphore(rifsc_base +
+							_RIFSC_RISC_PER0_SEMCR +
+							per_offset,
+							MAX_CID_SUPPORTED))
+				return -1;
+
+			sem_taken = true;
+		}
 	}
 
 	if (stm32_mux_set_parent(mux, sel))
-- 
2.47.2

